source/mir/ndslice/sorting.d

```diff
@@ -676,3 +676,212 @@ version(mir_test)
 
     assert(arr.indexed(index).pairwise!"a < b".all);
 }
+
+/++
+Default function for partitionAt
+
+Params:
+    slice = input 1-dimensional slice
++/
+@trusted pure @nogc nothrow
+template setPivotAt(alias less) {
+    size_t setPivotAt(Iterator, SliceKind kind)
+        (Slice!(Iterator, 1, kind) slice)
+    {
+        size_t len = slice.length;
+
+        auto leftI = slice._iterator;
+        auto midI = leftI + len / 2;
+        auto rightI = leftI + len - 1;
+
+        setPivot!less(len, leftI, midI, rightI);
+        return midI - leftI;
+    }
+}
+
+version(mir_test)
+@safe pure
+unittest {
+    import mir.functional: naryFun;
+    auto x = [3, 1, 5, 2, 0].sliced;
+
+    auto y = setPivotAt!(naryFun!("a < b"))(x);
+    assert(y == 2);
+    assert(x[0 + y] == 5);
+
+    auto z = setPivotAt!(naryFun!("a < b"))(x[1..3]);
+    assert(z == 1);
+    assert(x[1 + z] == 5);
+}
+
+/++
+Partitions a `slice` at a value `k`, such that for each `x` in `slice[0..k]` the 
+function `less(x, slice[k])` is true or `x` equals `slice[k]` and for each `y`  
+in `slice[k..$]` the function `less(slice[k], y)` is true or `y` equals 
+`slice[k]`.
+
+The function returns the value `slice[k]`, which is equivalent to 
+returning the value of `slice[k]` if slice is sorted with `sort!less`.
+
+Can also provide a pivotFunction that is used to choose pivot points in the 
+implementation. The pivotFunction must conform to the following signatures:
+    size_t value = pivotFunction!less(slice, left, right);
+    size_t value = pivotFunction(slice, left, right);
+
+Params:
+    slice = input 1-dimensional slice
+    k = partition value
+Returns: 
+    the value slice[k]
++/
+@safe pure @nogc nothrow
+template partitionAt(alias less = "a < b", 
+                     alias pivotFunction = setPivotAt)
+{
+    import mir.functional: naryFun;
+
+    static if (__traits(isSame, naryFun!less, less))
+    {
+        DeepElementType!(Slice!(Iterator, 1, kind))
+            partitionAt
+            (Iterator, SliceKind kind)(
+                Slice!(Iterator, 1, kind) slice, 
+                size_t k) 
+        {
+            assert(slice.length > 0, 
+                "partitionAt: slice must have length greater than 0");
+            assert(k >= 0, 
+                "partitionAt: k must be greater than or equal to zero");
+            assert(k < slice.length, 
+                "partitionAt: k must be less than the length of the slice");
+
+            if (slice.length == 1) {
+                return slice[0];
+            }
+
+            import std.algorithm.sorting: pivotPartition;
+
+            size_t left = 0;
+            size_t right = slice.length;
```

Comment:
Instead of having two additional variables, we can just adjust `k` and `slice` itself. It would be much either to review and a bit faster